jupyter_client/blocking/client.py

```diff
@@ -74,3 +132,179 @@ def wait_for_ready(self, timeout=None):
+        """
+        if reply and not self.iopub_channel.is_alive():
+            raise RuntimeError("IOPub channel must be running to receive output")
+        if allow_stdin is None:
+            allow_stdin = self.allow_stdin
+        if allow_stdin and not self.stdin_channel.is_alive():
+            raise RuntimeError("stdin channel must be running to allow input")
+        msg_id = super(BlockingKernelClient, self).execute(code,
+                              silent=silent,
+                              store_history=store_history,
+                              user_expressions=user_expressions,
+                              allow_stdin=allow_stdin,
+                              stop_on_error=stop_on_error,
+        )
+        if not reply:
+            return msg_id
+
+        if 'IPython' in sys.modules:
+            from IPython import get_ipython
+            ip = get_ipython()
+            in_kernel = getattr(ip, 'kernel', False)
+            if in_kernel:
+                socket = ip.display_pub.pub_socket
+                session = ip.display_pub.session
+                parent_header = ip.display_pub.parent_header
+        else:
+            in_kernel = False
+
+        # set deadline based on timeout
+        if timeout is not None:
+            deadline = monotonic() + timeout
+        
+        poller = zmq.Poller()
+        iopub_socket = self.iopub_channel.socket
+        poller.register(iopub_socket, zmq.POLLIN)
+        if allow_stdin:
+            stdin_socket = self.stdin_channel.socket
+            poller.register(stdin_socket, zmq.POLLIN)
+        else:
+            stdin_socket = None
+
+        # wait for output and redisplay it
+        while True:
+            if timeout is not None:
+                timeout = max(0, deadline - monotonic())
+            events = dict(poller.poll(timeout=timeout))
+            if not events:
+                raise TimeoutError("Timeout waiting for IPython output")
+            if stdin_socket in events:
+                req = self.stdin_channel.get_msg(timeout=0)
+                self._handle_input(req)
+                continue
+            if iopub_socket not in events:
+                continue
+
+            msg = self.iopub_channel.get_msg(timeout=0)
+
+            if msg['parent_header'].get('msg_id') != msg_id:
+                # not from my request
+                continue
+            msg_type = msg['header']['msg_type']
+            content = msg['content']
+            if msg_type == 'status':
+                if content['execution_state'] == 'idle':
+                    # idle means output is done
+                    break
+            elif msg_type == 'stream':
+                stream = getattr(sys, content['name'])
+                stream.write(content['text'])
+            elif msg_type in ('display_data', 'execute_result', 'error'):
+                if in_kernel:
+                    session.send(socket, msg_type, content, parent=parent_header)
```

Comment:
Sounds good to me. I'll leave `execute(reply=True)` matching the other methods, without touching output.